packages/tokamak/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol

```diff
@@ -54,4 +79,141 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
+        }
+
+        // We use the v1 message hash as the unique identifier for the message because it commits
+        // to the value and minimum gas limit of the message.
+        bytes32 versionedHash =
+            Hashing.hashCrossDomainMessageV1(_nonce, _sender, _target, _value, _minGasLimit, _message);
+
+        if (_isOtherMessenger()) {
+            // These properties should always hold when the message is first submitted (as
+            // opposed to being replayed).
+            assert(msg.value == 0);
+            assert(!failedMessages[versionedHash]);
+        } else {
+            require(msg.value == 0, "CrossDomainMessenger: value must be zero unless message is from a system address");
+
+            require(failedMessages[versionedHash], "CrossDomainMessenger: message cannot be replayed");
+        }
+
+        require(
+            _isUnsafeTarget(_target) == false, "CrossDomainMessenger: cannot send message to blocked system address"
+        );
+
+        require(successfulMessages[versionedHash] == false, "CrossDomainMessenger: message has already been relayed");
+
+        // If there is not enough gas left to perform the external call and finish the execution,
+        // return early and assign the message to the failedMessages mapping.
+        // We are asserting that we have enough gas to:
+        // 1. Call the target contract (_minGasLimit + RELAY_CALL_OVERHEAD + RELAY_GAS_CHECK_BUFFER)
+        //   1.a. The RELAY_CALL_OVERHEAD is included in `hasMinGas`.
+        // 2. Finish the execution after the external call (RELAY_RESERVED_GAS).
+        //
+        // If `xDomainMsgSender` is not the default L2 sender, this function
+        // is being re-entered. This marks the message as failed to allow it to be replayed.
+        if (
+            !SafeCall.hasMinGas(_minGasLimit, RELAY_RESERVED_GAS + RELAY_GAS_CHECK_BUFFER)
+                || xDomainMsgSender != Constants.DEFAULT_L2_SENDER
+        ) {
+            failedMessages[versionedHash] = true;
+            emit FailedRelayedMessage(versionedHash);
+
+            // Revert in this case if the transaction was triggered by the estimation address. This
+            // should only be possible during gas estimation or we have bigger problems. Reverting
+            // here will make the behavior of gas estimation change such that the gas limit
+            // computed will be the amount required to relay the message, even if that amount is
+            // greater than the minimum gas limit specified by the user.
+            if (tx.origin == Constants.ESTIMATION_ADDRESS) {
+                revert("CrossDomainMessenger: failed to relay message");
+            }
+
+            return;
+        }
+
+        xDomainMsgSender = _sender;
+        bool transferStatus = true;
+        if (_value != 0){
+            transferStatus = IERC20(l1TONAddress).transfer(_target, _value);
```

Comment:
@rlgns98kr I updated approve to transfer. The function safeTransfer does not return tx's status, and in our case,  the function transfer in TON returns tx's status so that it is safe to use transfer instead of safeTransfer.

Reply:
Thank you very much! 🙇🏼‍♂️

Reply:
@rlgns98kr I think we should re-consider this case. Because as I see, It seems like the normal logic of smart contracts when receiving ERC-20 token (to process arbitrary logic) is using transferFrom. For example, when we use our Bridge or Uniswap, we need to approve first and then calling swap functions. 

What do you think about that?

Reply:
That sounds very reasonable. But I think the context is a little different here. This part is when we pass the token to the contract we will call.

(In Uniswap, if we assume that the user transfers ERC20 directly to Uniswap, the process will be very complex.)

As you know, if we decide to only approve here, the users must check whether `transferFrom` is exist in every message they relays (a function to be executed in L1) in order to receive an ERC20 token. But if this is common process in the Contract world, I strongly agree with you.

In conclusion, since we have no experience with contracts, it would be a good idea to explain this problem to Zena or Harvey and ask for their opinions.

Thank you so much for checking again in detail.
If you agree, I'll tag them here. 🙇🏼‍♂️

Reply:
Thank you very much, Steven
It is great if having their opinions

Reply:
As I understood, if we use transfer() and then call smart contract logic, smart contracts dont know how many token that sender transfered for the call, so that ERC20 standard has approve() and transferFrom() to solve that situation.

Reply:
Oh,, It's quite difficult. The target contract have to check if erc20 is transferred! Okay. Let's use approve. I'm sorry for my lack of experience and knowledge about Contract,,,

Reply:
Do you think we should have contract team's advise for sure?

Reply:
And I think we have to consider several security problems in here. If you are okay, we can tag here Zena and Harvey in here! https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20-approve-address-uint256-
🙇🏼‍♂️

Reply:
Ah, I did mean I am okay.

Reply:
@Zena-park @zzooppii 
Could you please take a look at that part when it's convenient for you?
This is the part where tx started from L2 is executed in L1. Since TON will be used as a native token in L2, ERC20(TON) in L1 must be passed to the target contract before execution. Is there anything that needs to be added or considered?

Reply:
Sorry for the very late feedback.

Reply:
Sorry for the very late feedback.

L1's [TON](https://github.com/tokamak-network/plasma-evm-contracts/blob/master/contracts/stake/tokens/TON.sol) has [approveAndCall](https://github.com/tokamak-network/plasma-evm-contracts/blob/858b2040b67d7de1f7cfa3b477e62dc938ca4499/contracts/stake/tokens/ERC20OnApprove.sol#L10-L14) function. 
Therefore, if L1StandardBridge supports the onApprove function (e.g. [WTON](https://github.com/tokamak-network/plasma-evm-contracts/blob/858b2040b67d7de1f7cfa3b477e62dc938ca4499/contracts/stake/tokens/WTON.sol#L44-L49)), it will be convenient as customers can use it without prior approval of TON.